#  HELPING  **ETS KATS Trucks** WITH CAR SALES DATA 

## 1.1  Introduction 
**ETS KATS Trucks**  is a Cameroonian based company specialized in the purchase of automotive equipments ranging from cars, trucks, construction equipments and spare parts from all European brands. This company is new to the market and needs to ubderstand how the market functions.

We know just little about the  competitors in the market, but we are aware that ***eBay Kleinanzeigen***, the car used section of the German eBay website, is fairly popular, and extracting the data from it is easier than from other private selling competitors.

After some carefull research, a data set was found [here](https://data.world/data-society/used-cars-data), containing 50,000 data points. This data set will be analysed to provide useful insights for the newbie com
pany.


* Importing Necessary liabraries

In [76]:
import numpy as np
import pandas as pd

* The data found is in csv format hence has little readability. To get this data visaully appealing for analysis, it will be read into a dataframe (tabular format).

In [77]:
autos = pd.read_csv("autos.csv",encoding = "Latin-1")

#Data is read and stored in a variable called autos.
#The encoding used here is "Latin-1". This is because the default "UTF-8" format gave us an error

It is important to explore what kind of information is contained in our newly created `autos` dataframe

In [78]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [79]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## 1.2  Initial Observation.

**We have 20 columns and 50.000 rows. At a first glance we can observe several things that might affect our analysis:**
1. `Price` column has type as `object` which is improper.
2. There are several models,  fuel type, gearbox types and vehicle types missing. This could indicate, perhaps, that there are some vehicle parts also listed
3. Missing values in fuel type could mean vehicles being sold without engine
4. We also need to be careful with the missing values of notRepairedDamage. Does this mean that there is no damage, or that if there is damage is not repaired?


## 1.3 Format and readability.
The course of analysis will ensure constant interaction with the data and so, clear formating for naming columns should be implemented.

At this point, the column names will be properly formatted to reflect the ***snakecase*** (using underscore inbetween words to represent spaces) instead of the presented ***camelCase***(using capital letters to indicate the beginning of new word). This is to make this data ready for analysis and to follow standard convention.

In [80]:
autos.rename({"dateCrawled" : "date_crawled"}, axis = 1, inplace = True)
autos.rename({"offerType" : "offer_type"}, axis = 1, inplace = True)
autos.rename({"vehicleType" : "vehicle_type"}, axis = 1, inplace = True)
autos.rename({"yearOfRegistration" : "registration_year"}, axis = 1, inplace = True)
autos.rename({"powerPS" : "power_ps"}, axis = 1, inplace = True)
autos.rename({"monthOfRegistration" : "registration_month"}, axis = 1, inplace = True)
autos.rename({"fuelType" : "fuel_type"}, axis = 1, inplace = True)
autos.rename({"notRepairedDamage" : "unrepaired_damage"}, axis = 1, inplace = True)
autos.rename({"dateCreated" : "ad_created"}, axis = 1, inplace = True)
autos.rename({"nrOfPictures" : "nr_pictures"}, axis = 1, inplace = True)
autos.rename({"postalCode" : "postal_code"}, axis = 1, inplace = True)
autos.rename({"lastSeen" : "last_seen"}, axis = 1, inplace = True)

* Examining to seechanges in the column names.

In [81]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

* Of the most stunning formating done to enhance readability include the following:
    1. `yearOfRegistration` changed to `registration_year`
    2. `monthOfRegistration` replaced by `registration_month`
    3. `notRepairedDamage`converted to `unrepaired_damage`

## 1.4 Further Exploration - Observation
Sofar, we have used the `info()` and the `head()` methods to have a glance at our dataframe. To better investigate what obnormalities our dataset presents, we are going to be making use of the `dataframe.describe()` to obtainm descriptive statistics.

In [82]:
autos.describe(include="all")

# include = "all" ensures that all column types, number and string alike should be included.

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_pictures   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

#### 1.4.1 Non Numeric Values
We see `price` and `odometer` have NaN values in the numeric statistics, this is because the rather have `string` type instead of `number` as we saw ealier. If we look up the autos.

This could mean that price has currency symbols in its values, and the odometer kilometer values.

Let's see:

In [83]:
autos["price"].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [84]:
autos["odometer"].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object

* It is clearly seen that they have symbols in their values. This needs to be treated.

##### 1.4.2 Odd value

* Taking a look at the `registration_year` column, it is seen that the minimum year **1000** is too far into the past to be considered validd. Cars were not even know around as the first car was rather produced in **January 29, 1886**. Also, the maximum year **9999** is many decades into the future. This clearly shows that there are values here that are out of the expected range.

* The `power_PS` ps data has a very similar issue, **17,700 ps (horse power)** too high even as a quick google search reveals that A4 Cabriolet has a horse power of just 220 PS.

## 2. Data Cleaning

### 2.1 "price" and "odometer" columns

#### 2.1.1 Transforming values to numbers

As ealier seen, `price` and `odometer` columns both contain  values of type `object` instead of `float64`. This is because they have non-numeric symbols which all need to be gotten rid of.

In [85]:
# Removing non numeric characters from `price` and converting it to `numeric`
autos["price"] = autos["price"].str.replace("$", "", regex=False)
autos["price"] = autos["price"].str.replace(",", "")
autos["price"] = autos["price"].astype(float)

* Inspecting the Effect of these changes.

In [87]:
autos["price"].head()

0    5000.0
1    8500.0
2    8990.0
3    4350.0
4    1350.0
Name: price, dtype: float64

In [74]:

#Removing non numeric characters from `odometer` and converting it to `numeric`
# autos["odometer"] = autos["odometer"].str.replace(",","").str.replace("km","").astype(float)
autos["odometer_km"] = autos["odometer"].str.replace("km", "")
autos["odometer_km"] = autos["odometer_km"].str.replace(",", "")
autos["odometer_km"] = autos["odometer_km"].astype(float)


In [ ]:
autos["odometer_km"].head()

#  Exploring the Odometer and Price Columns

* Let us to some further anakysis on the **odometer and price** columns to see if there are any outliers worth removing

In [ ]:
# low, high = autos["price"].quantile([0.1,0.9])
# low1,high1 = autos["odometer_km"].quantile([0.1,0.9])
print(autos["price"].unique().shape)

In [ ]:
print(autos["price"].value_counts().head())
print("------")
print(autos["price"].value_counts().sort_index(ascending = False).head(20))

In [ ]:
low, high = autos["price"].quantile([0.1, 0.9])
autos.loc[:,"price"] = autos[autos["price"].between(low,high)]

It is observerved here that the `price` column has **zero** as being the minimum which is unrealistic. Also the higest value is about


We resolve outliers by removing the the bottom 

In [ ]:
autos[["price"]].info()

In [ ]:
print(autos["odometer_km"].value_counts().sort_index())

No outliers for the `odometer` hence it should be left untouched.

# Exploring the date columns

We get the distribution of values in the `date_crawled and last_seen  `

In [ ]:
# Date distribution for `date_crawled` 
autos["date_crawled"].str[:10].value_counts(dropna=False, normalize = True).sort_index()

In [ ]:
# Date distribution for `last_seen `
autos["last_seen"].str[:10].value_counts(dropna=False, normalize = True).sort_index()

Exploring the two series shows that this data collected in **2016** only include the months of **March** and **April**

In [ ]:
autos["registration_year"].describe()
# autos.columns

It is  observed that the minimum value is **1000** which is far back in the past even before cares were created. As well, the maximum value states **9999** which is many years into the future. This clearly shows that there are outliers in our data set.

So, let us check for the values that seperate the bottom and the top 10 percent of our data set.

In [ ]:
# low, high = autos["registration_year"].quantile([0.001,0.95])

In [ ]:
autos["registration_year"].value_counts().sort_index().head(40)

In [ ]:
autos["registration_year"].value_counts().sort_index().tail(30)

Observing the year values from the `head` shows a drastic jump from  **1800** to **1910**,also, the number of counts are very limited until **1964**. Oberving from the `tail` shows an abnormal gap between **2019** and **2028** but reasonable count of car registration only ends in **2018**

So, we now filter this data to contain only car registration between between the years **1964** and **2018** such that any any registration year out of this range will simply be replaced by `NAN`

In [ ]:
autos.loc[:,"registration_year"] = autos[autos["registration_year"].between(1964,2018)]

In [ ]:
autos["registration_year"].describe()

It is immediately noticed that the problem ealier mentioned has been removed as justified by the the descriptive statistics above.

# Exploring Price by Brand

In [ ]:
autos["brand"].unique()

In [ ]:
top_car_brands = autos["brand"].value_counts(normalize=True)[:6]

In [ ]:
top_car_brands

As for the brand, *volkswagen* makes up *21%*, *opel* takes *11%*, *bmw* occupies *11%* as well, *mercedes_benz* takes up *9%*, *audi* has *9%* whereas *ford* has *7%* of the car sales on eBay.

In [ ]:
brand_mean_prce = {}

for brand in top_car_brands.index:
    selected_rows = autos[autos["brand"] == brand]
    brand_price = selected_rows["price"].mean()
    brand_mean_prce[brand] = int(brand_price)
    
print(brand_mean_prce)

On average top cars and their cost is shown below

|car|mean_price|
|---|------|
|volkswagen|4,127|
|opel|2,984|
|bmw|5,340|
|mercedes_benz|4,860|
|audi|5,282|
|ford|3,196|

# Storing Aggregate Data in a DataFrame

* Here, we aggregate the the price and odometer of the cars per brand. The purpose of doing this is to see if there exist any relationship between mileage and the mean price of the top brand cars available on eBay

In [ ]:
brand_mean_prce = pd.Series(brand_mean_prce).sort_values(ascending = False)
pd.DataFrame(brand_mean_prce,columns=["mean_price"])

In [ ]:
brand_mean_mileage = {}

for brand in top_car_brands.index:
    selected_rows = autos[autos["brand"] == brand]
    brand_age = selected_rows["odometer_km"].mean()
    brand_mean_mileage[brand] = int(brand_age)
print(brand_mean_mileage)

In [ ]:
brand_mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending = False)
brand_information = pd.DataFrame(brand_mean_mileage,columns=["mean_mileage"])
brand_information

In [ ]:
# pd.DataFrame(brand_mean_prce,brand_mean_mileage, columns= ["mean_price","mile_age"])

brand_information["mean_price"] = brand_mean_prce

brand_information

This display clearly shows that mean mileage varies very less with respect to the mean price for the top car brands.